### ISSUE TITLE
Figure out why the attached tf_gather.py TensorFlow script is slow going on with this tf_gather.py script

### ISSUE THREAD

**Issue by cfregly** [fluxcapacitor/pipeline#119]
From a community partner:

"Apparently TF is rather slow in reductions and mutliplications of vectors. Run this example and you will see very low CPU usage. I know I can run this with sparse vectors. It is probably optimized. I am disappointed this didn't work."

[tf_gather.py.zip](https://github.com/fluxcapacitor/pipeline/files/451083/tf_gather.py.zip)

**Comment by cfregly** [comment]
comments from @samjabrahams:

- When you put `sess.run()` inside a loop, you're having TensorFlow jump between the Python and C++ layers 100000 times, which just exacerbates the communication bottleneck. In typical TensorFlow code, you end up with a large amount of operations being run with a single call to `run()`

- I see TensorFlow using up all of a single core with the provided code. The code is iterating through and running a bunch of simple, single threaded operations sequentially, so that's why you don't see large overall CPU usage
 
- I've attached some modified code: it places 100000 multiplies to be run in sequence in a single graph and then runs the whole thing with a single call to `sess.run()`. It doesn't use multiple cores, however. On my computer, the script finishes running in 828.747865915, or approximately 0.0083 seconds per series of operations created with `func()`
 
- Check out the discussion here on the GitHub repo about creating code to test out multiple-CPU usage. The basic idea is to create a TensorFlow Queue along with helpful built-in functionality to create a bunch of coordinated threads to run the code.

[one_session_call.py.zip](https://github.com/fluxcapacitor/pipeline/files/452336/one_session_call.py.zip)

**Comment by cfregly** [comment]
response from community member:

"It is a problem that I have seen in TF, in general. I was discussing it with a friend from <redacted>. TF is very good at dense linear algebra. In fact if you run the same code with a matrix product or a hadamard product you will see very good CPU utilization. 

Try to increase the size of this example and you would expect that most of the time will be spent on the reduce_sum. It doesn't happen. Switching and serialization still dominate. I am not sure that it actually does parallelize the reduce. 

In fact with Sam's way it just executes all the nodes in parallel but not each node. In general I think that TF parallelize only linear algebra nodes. Also sometimes it can parallelize in GPU but not in GPU.

The bottom line is that if you want to run things fast it is not as simple as you think. You need to get deep down to control flow or use the trick with the queue which is not really a great paradigm. Also pushing all the data in the graph is not a wise thing as you copy a lot."
